# Statistical Methods in Pandas - Lab

## Introduction

In this lesson you'll get some hands on experience using some of the key summary statistics methods in Pandas.

## Objectives:
You will be able to:

* Understand and use the df.describe() and df.info() summary statistics methods
* Use built-in Pandas methods for calculating summary statistics (.mean(), .std(), .count(), .sum(), .mean(), .median(), .std(), .var() and .quantile())
* Apply a function to every element in a Series or DataFrame using s.apply() and df.applymap()


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets.  You will find this dataset in the file `lego_sets.csv`.  

In the cell below:

* Import pandas and set the standard alias of `pd`
* Load in the `lego_sets.csv`dataset using the `read_csv()` function
* Display the head of the DataFrame to get a feel for what we'll be working with

In [1]:
import pandas as pd

In [4]:
df = pd.read_csv('lego_sets.csv')
df.head()

ages  list_price  num_reviews  piece_count  play_star_rating  \
0  6-12       29.99          2.0        277.0               4.0   
1  6-12       19.99          2.0        168.0               4.0   
2  6-12       12.99         11.0         74.0               4.3   
3   12+       99.99         23.0       1032.0               3.6   
4   12+       79.99         14.0        744.0               3.2   

                                           prod_desc  prod_id  \
0  Catapult into action and take back the eggs fr...  75823.0   
1  Launch a flying attack and rescue the eggs fro...  75822.0   
2  Chase the piggy with lightning-fast Chuck and ...  75821.0   
3  Explore the architecture of the United States ...  21030.0   
4  Recreate the Solomon R. Guggenheim Museum® wit...  21035.0   

                                      prod_long_desc review_difficulty  \
0  Use the staircase catapult to launch Red into ...           Average   
1  Pilot Pig has taken off from Bird Island with ...              Easy   
2  Pitch speedy bird Chuck against the Piggy Car....              Easy   
3  Discover the architectural secrets of the icon...           Average   
4  Discover the architectural secrets of Frank Ll...       Challenging   

                         set_name  star_rating    theme_name  val_star_rating  \
0           Bird Island Egg Heist          4.5  Angry Birds™              4.0   
1              Piggy Plane Attack          5.0  Angry Birds™              4.0   
2                Piggy Car Escape          4.3  Angry Birds™              4.1   
3  United States Capitol Building          4.6  Architecture              4.3   
4   Solomon R. Guggenheim Museum®          4.6  Architecture              4.1   

  country  
0      US  
1      US  
2      US  
3      US  
4      US

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset.  There are two ways we'll get this information-- `.info()` and `.describe()`.

### Using `.info()`

The `.info()` method provides us metadata on the DataFrame itself.  This allows to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset? 
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method. 

In [13]:
df.info()
df[df.isnull().any(1)]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
ages                 12261 non-null object
list_price           12261 non-null float64
num_reviews          10641 non-null float64
piece_count          12261 non-null float64
play_star_rating     10486 non-null float64
prod_desc            11884 non-null object
prod_id              12261 non-null float64
prod_long_desc       12261 non-null object
review_difficulty    10206 non-null object
set_name             12261 non-null object
star_rating          10641 non-null float64
theme_name           12258 non-null object
val_star_rating      10466 non-null float64
country              12261 non-null object
dtypes: float64(7), object(7)
memory usage: 1.3+ MB


ages  list_price  num_reviews  piece_count  play_star_rating  \
22       10+      9.9900          1.0        136.0               NaN   
32       10+     19.9900          NaN        209.0               NaN   
48      5-12     49.9900          NaN        387.0               NaN   
55      7-12     99.9900          1.0        883.0               5.0   
69      5-12     39.9900          NaN        297.0               NaN   
84       4-7     19.9900          NaN        115.0               NaN   
92       4-7      9.9900          NaN         73.0               NaN   
94      5-12      9.9900          2.0         88.0               NaN   
103     5-12     49.9900          2.0        403.0               4.5   
108     5-12     29.9900          NaN        195.0               NaN   
153     6-12      9.9900          NaN        109.0               NaN   
162     7-12     19.9900          NaN        246.0               NaN   
163     6-12      9.9900          2.0        124.0               4.0   
192     7-14     19.9900          2.0        198.0               4.0   
199      10+     19.9900          NaN        209.0               NaN   
205     9-12     99.9900          NaN       1078.0               NaN   
208     7-12     59.9900          NaN        432.0               NaN   
227     9-12     99.9900          NaN       1078.0               NaN   
228     7-12     59.9900          NaN        432.0               NaN   
231     7-14     14.9900          NaN         56.0               NaN   
234     7-14     24.9900          NaN         93.0               NaN   
261     5-12     19.9900          NaN        125.0               NaN   
267      4-7      9.9900          1.0         75.0               5.0   
277     5-12      9.9900          NaN         75.0               NaN   
278     5-12      5.9900          NaN         61.0               NaN   
279      2-5     29.9900          NaN         22.0               NaN   
280      2-5     19.9900          1.0         24.0               NaN   
284      2-5     59.9900          NaN        104.0               NaN   
285      2-5     59.9900          NaN         98.0               NaN   
287      2-5     49.9900          NaN         71.0               NaN   
...      ...         ...          ...          ...               ...   
12048  10-21    243.9878          1.0          1.0               5.0   
12049  10-21     42.6878          NaN          1.0               NaN   
12050    10+     42.6878          4.0          1.0               3.0   
12051  10-21     42.6878          1.0          1.0               5.0   
12052  10-21     42.6878          NaN          1.0               NaN   
12053    10+     42.6878          3.0          1.0               5.0   
12054     8+     36.5878          2.0          1.0               5.0   
12055  10-21     36.5878          2.0          1.0               5.0   
12056  10-21     30.4878          1.0          1.0               5.0   
12057  10-21     24.3878          NaN          1.0               NaN   
12058  10-21     18.2878          NaN          7.0               NaN   
12065     8+     54.8878          NaN        278.0               NaN   
12071   6-14     18.2878          NaN         84.0               NaN   
12078   8-14     69.5278          NaN        569.0               NaN   
12080   8-14     36.5878          NaN        369.0               NaN   
12082   7-14     24.3878          NaN        217.0               NaN   
12096   8-14     54.8878          NaN        530.0               NaN   
12106    4-7     24.3878          NaN        131.0               NaN   
12108   6-14     12.1878          NaN         61.0               NaN   
12111   6-14     12.1878          NaN         69.0               NaN   
12113   6-14     12.1878          NaN         69.0               NaN   
12115     6+    853.9878          7.0       2631.0               4.7   
12116     6+    683.1878          8.0       2455.0               3.7   
12117     6+    439.1878          2.0       4900

#### Interpreting the Results

Read the output above, and then answer the following questions:

How many total rows are in this DataFrame?  How many columns contain numeric data? How many contain categorical data?  Identify at least 3 columns that contain missing values. 

Write your answer below this line:
________________________________________________________________________________________________________________________________

rows = 12261
numeric cols = 7
categorical = 7 


Answer question here...

## Using `.describe()`

Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method. 

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns.  Use these to answer the following questions:

How much is the standard deviation for piece count?  How many pieces are in the largest lego set?  How many in the smallest lego set? What is the median `val_star_rating`?

________________________________________________________________________________________________________________________________

Answer questions here...

## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods. 

In the cell below, compute the median value of the `star_rating` column.

Next, get a count of the total number of values in `play_star_rating`.

Now, compute the standard deviation of the `list_price` column.

If we bought every single lego set in this dataset, how many pieces would we have?  Use the `.sum()` method on the correct column to compute this. 

Now, let's try getting the value for the 90% quantile.  Do this in the cell below.

## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data--there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data. 

In the cell below, get the `.unique()` values contained within the `review_difficulty` column. 

Now, let's get the `value_counts` for this column, to see how common each is. 

As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `applymap()` function and passing in a lambda function. 

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call our DataFrame's `.applymap()` function and pass in `lambda x: str(x)`
* Call our new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [ ]:
string_df = None

Note that everything--even the `NaN` values, have been cast to a string in the example above. 

Note that for pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame--this will become especially useful when we learn how to **_normalize_** our datasets in a later section!

# Conclusion

In this lab, we learned how to:

* Understand and use the df.describe() and df.info() summary statistics methods
* Use built-in Pandas methods for calculating summary statistics (.mean(), .std(), .count(), .sum(), .mean(), .median(), .std(), .var() and .quantile())
* Apply a function to every element in a Series or DataFrame using s.apply() and df.applymap()